PurchasesCoreSwift/Networking/HTTPClient.swift

```diff
@@ -0,0 +1,314 @@
+                Logger.debug(logMessage)
+                self.currentSerialRequest = nil
+                if !self.queuedRequests.isEmpty {
+                    maybeNextRequest = self.queuedRequests[0]
+                    self.queuedRequests.remove(at: 0)
+                }
+                if let nextRequest = maybeNextRequest {
+                    Logger.debug(String(format: Strings.network.starting_next_request, nextRequest))
+                    self.performRequest(nextRequest.httpMethod,
+                                        performSerially: true,
+                                        path: nextRequest.path,
+                                        requestBody: nextRequest.requestBody,
+                                        headers: nextRequest.headers,
+                                        completionHandler: nextRequest.completionHandler)
+                }
+            }
+        }
+    }
+
+    func createRequest(httpMethod: String,
+                       path: String,
+                       requestBody maybeRequestBody: [AnyHashable: Any]?,
+                       headers: [String: String],
+                       refreshETag: Bool) -> URLRequest? {
+        let relativeURLString = "/v1\(path)"
+        guard let requestURL = URL(string: relativeURLString, relativeTo: SystemInfo.serverHostURL) else {
+            return nil
+        }
+
+        var request = URLRequest(url: requestURL)
+        request.httpMethod = httpMethod
+
+        let eTagHeader = eTagManager.eTagHeader(for: request, refreshETag: refreshETag)
+        let headersWithETag = headers.merging(eTagHeader) { (_, last) -> String in
+            last
+        }
+
+        request.allHTTPHeaderFields = headersWithETag
+
+        if httpMethod == "POST" {
+            if let requestBody = maybeRequestBody {
+                var jsonParseError: Error?
+                let isValidJSONObject = JSONSerialization.isValidJSONObject(requestBody)
+                if isValidJSONObject {
+                    do {
+                        request.httpBody =
+                            try JSONSerialization.data(withJSONObject: requestBody)
+                    } catch let error {
+                        jsonParseError = error
+                    }
+                }
+                if !isValidJSONObject || jsonParseError != nil {
+                    Logger.error(String(format: Strings.network.creating_json_error, requestBody))
+                    return nil
+                }
+            }
+        }
+        return request
+    }
+
+    func assertIsValidRequest(httpMethod: String, requestBody: [String: Any]?) throws {
+        if (httpMethod != "GET" && httpMethod != "POST") ||
+            (httpMethod == "GET" && requestBody != nil) ||
+            (httpMethod == "POST" && requestBody == nil) {
+            throw HTTPClientError.invalidNetworkCall(httpMethod, requestBody: requestBody)
+        }
+    }
```

Comment:
not needed for this PR, but I really think we should be able to clean this up nicely by having a different entry point for post and get, where one allows for a body.
we don't use other http methods for now. 
if we ever do, we could refactor into an enum. 

That way validation would be unnecessary since we'd be enforcing format at the method signature level.

Reply:
It would simplify a lot the performRequest function too since body for example wouldn't be nil for POSTS and we would ignore it for GETs

Reply:
@aboedo would you mind opening an issue with this suggestion? I think it would be great to keep track of suggestions like this!